lib/ansible/modules/network/avi/avi_api_session.py

```diff
@@ -0,0 +1,187 @@
+#!/usr/bin/python
+"""
+# Created on Aug 12, 2016
+#
+# @author: Gaurav Rastogi (grastogi@avinetworks.com) GitHub ID: grastogi23
+#
+# module_check: not supported
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+
+import json
+import time
+from ansible.module_utils.basic import AnsibleModule
+from copy import deepcopy
+
+HAS_AVI = True
+try:
+    from avi.sdk.avi_api import ApiSession, ObjectNotFound
+    from avi.sdk.utils.ansible_utils import (
+        ansible_return, avi_obj_cmp, cleanup_absent_fields,
+        avi_common_argument_spec)
+except ImportError:
+    HAS_AVI = False
+
+
+DOCUMENTATION = '''
+---
+module: avi_api
+author: Gaurav Rastogi (grastogi@avinetworks.com)
+
+short_description: API Module for invoking any Avi API.
+description:
+    - This module is used to configure Pool object
+version_added: 2.3
+requirements: [ avisdk ]
+options:
+    http_method:
+        description:
+            - Allowed HTTP methods for RESTful services and are supported by Avi Controller.
+        choices: ["get", "put", "post", "patch", "delete"]
+        required: true
+    data:
+        description:
+            - HTTP body in YAML format.
+    data_json:
+        description:
+            - HTTP body in JSON format.
+    params:
+        description:
+            - Query parameters passed to the HTTP API.
+extends_documentation_fragment:
+    - avi
+'''
+
+EXAMPLES = '''
+# Get Pool Information using avi_api_session
+- avi_api_session:
+    # get pool information
+    controller: "{{ controller }}"
+    username: "{{ username }}"
+    password: "{{ password }}"
+    http_method: get
+    path: pool
+    params:
+      name: "{{ pool_name }}"
+  register: pool_results
+
+# Patch Pool with list of servers
+  - avi_api_session:
+      controller: "{{ controller }}"
+      username: "{{ username }}"
+      password: "{{ password }}"
+      http_method: patch
+      path: "{{ pool_path }}"
+      data:
+        add:
+          servers:
+            - ip:
+                addr: 10.10.10.10
+                type: V4
+            - ip:
+                addr: 20.20.20.20
+                type: V4
+'''
+
+
+RETURN = '''
+obj:
+    description: Avi REST resource
+    returned: success, changed
+    type: dict
```

Comment:
and this shows problem with such generic module, user MUST know api to know what is returned